xarray/core/duck_array_ops.py

```diff
@@ -214,26 +214,32 @@ def astype(data, dtype, **kwargs):
 
 
 def asarray(data, xp=np):
+    print(data)
+    print(type(data))
     return data if is_duck_array(data) else xp.asarray(data)
 
 
-def as_shared_dtype(scalars_or_arrays, xp=np):
-    """Cast a arrays to a shared dtype using xarray's type promotion rules."""
-    array_type_cupy = array_type("cupy")
-    if array_type_cupy and any(
-        isinstance(x, array_type_cupy) for x in scalars_or_arrays
-    ):
-        import cupy as cp
-
-        arrays = [asarray(x, xp=cp) for x in scalars_or_arrays]
+def as_duck_array(data, xp=np):
+    if is_duck_array(data):
+        return data
+    elif hasattr(data, "get_duck_array"):
+        # must be a lazy indexing class wrapping a duck array
+        return data.get_duck_array()
     else:
-        arrays = [asarray(x, xp=xp) for x in scalars_or_arrays]
```

Comment:
Previously this `asarray` call would coerce to numpy unnecessarily, when all we really wanted was an array type that we could examine the `.dtype` attribute of.